code/game/area/areas.dm

```diff
@@ -224,24 +226,93 @@ GLOBAL_LIST_EMPTY(teleportlocs)
 			turfs += T
 		map_generator.generate_terrain(turfs, src)
 
-/area/proc/get_contained_turfs()
-	if(length(turfs_to_uncontain))
+/// Returns the highest zlevel that this area contains turfs for
+/area/proc/get_highest_zlevel()
+	for (var/area_zlevel in length(turfs_by_zlevel) to 1 step -1)
+		if (length(turfs_to_uncontain_by_zlevel) >= area_zlevel)
+			if (length(turfs_by_zlevel[area_zlevel]) - length(turfs_to_uncontain_by_zlevel[area_zlevel]) > 0)
+				return area_zlevel
+		else
+			if (length(turfs_by_zlevel[area_zlevel]))
+				return area_zlevel
+	return 0
+
+/// Returns a nested list of lists with all turfs split by zlevel.
+/// only zlevels with turfs are returned. The order of the list is not guaranteed.
+/area/proc/get_zlevel_turf_lists()
+	if(length(turfs_to_uncontain_by_zlevel))
 		cannonize_contained_turfs()
-	return contained_turfs
+
+	var/list/zlevel_turf_lists = list()
+
+	for (var/list/zlevel_turfs as anything in turfs_by_zlevel)
+		if (length(zlevel_turfs))
+			zlevel_turf_lists[++zlevel_turf_lists.len] = zlevel_turfs
+
+	return zlevel_turf_lists
+
+/// Returns a list with all turfs in this zlevel.
+/area/proc/get_turfs_by_zlevel(zlevel)
+	if (length(turfs_to_uncontain_by_zlevel) >= zlevel && length(turfs_to_uncontain_by_zlevel[zlevel]))
+		cannonize_contained_turfs_by_zlevel(zlevel)
+
+	if (length(turfs_by_zlevel) < zlevel)
+		return list()
+
+	return turfs_by_zlevel[zlevel]
+
+
+/// Merges a list containing all of the turfs zlevel lists from get_zlevel_turf_lists inside one list. Use get_zlevel_turf_lists() or get_turfs_by_zlevel() unless you need all the turfs in one list to avoid generating large lists
+/area/proc/get_turfs_from_all_zlevels()
+	. = list()
+	for (var/list/zlevel_turfs as anything in get_zlevel_turf_lists())
+		. += zlevel_turfs
 
 /// Ensures that the contained_turfs list properly represents the turfs actually inside us
-/area/proc/cannonize_contained_turfs()
+/area/proc/cannonize_contained_turfs_by_zlevel(zlevel_to_clean, _autoclean = TRUE)
```

Comment:
I do not like these args

Reply:
?

Reply:
mostly what the heck is _autoclean, args like that smell off to me

Reply:
Its an internal argument that nobody other than the one place that sets it should fuck with.

Goals:

I want code that only looks at one zlevel to only uncontain turfs from that zlevel. - so I make a proc to uncontain turfs from a given zlevel

I want the uncontained turfs code to not be copied across many procs, - so i make the code that uncontains from all zlevels just call the first proc in a loop.

I want the uncontained turfs list to not have empty zlevels past the highest it has turfs to uncontain in, to ensure i can just check the outer length to know if there is any work to be done without having to look at or care about the inner lists. - so i have the uncontain proc strip off any empty zlevels at the end of the list.

and finally, I don't want effort to be repeated needlessly. looping over a list in a proc that sometimes gets called by code also looping over that same list is how you get `O(n^2)`. - so when the proc to uncontain from all zlevels runs, it tells the inner zlevel proc not to bother cleaning up the list, as its just gonna `= list()` it at the end.

It's named as it is because code that is running the zlevel specific proc on all zlevels is the only situation in which it is ok to not auto prune the list of empty zlevels at the tail.

and empty zlevels at the tail of the list is something I want to avoid because I need to ensure nothing accidentally depends on an assumption that the outer list is sized to world.maxz. something I thought about doing to avoid needing sanity checks, but requiring every area be notified of a zlevel change when 99% of them are gonna use the info to add a 21th empty list since most areas are only gonna have turfs in zlevels 2-5 seems silly. so keeping it sized as small as possible ensures anything not checking/asserting the size is highly likely to runtime